Megaparsec/Printable.lean

```diff
@@ -0,0 +1,95 @@
+import YatimaStdLib
+import Straume.Iterator
+
+open Straume.Iterator
+
+namespace Megaparsec.Printable
+
+class Printable (β : Type u) where
+  -- Pretty-print non-empty list of tokens. This function is also used
+  -- to print single tokens (represented as singleton lists).
+  showTokens : NEList β → String
+
+  -- Return the number of characters that a non-empty list of tokens
+  -- spans. The default implementation is sufficient if every token spans
+  -- exactly 1 character.
+  tokensLength : NEList β → Nat := fun xs => xs.toList.length
+
+export Printable (showTokens tokensLength)
+
+--===========================================================--
+--========================= HELPERS =========================--
+--===========================================================--
+
+-- TODO: why are almost all escape sequences unrecognised by Lean?
+private def charPretty' : Char → Option String
+  -- | '\NUL' => .some "null"
+  -- | '\SOH' => .some "start of heading"
+  -- | '\STX' => .some "start of text"
+  -- | '\ETX' => .some "end of text"
+  -- | '\EOT' => .some "end of transmission"
+  -- | '\ENQ' => .some "enquiry"
+  -- | '\ACK' => .some "acknowledge"
+  -- | '\BEL' => .some "bell"
+  -- | '\BS' => .some "backspace"
+  | '\t' => .some "tab"
+  | '\n' => .some "newline"
+  -- | '\v' => .some "vertical tab"
+  -- | '\f' => .some "form feed"
+  | '\r' => .some "carriage return"
+  -- | '\SO' => .some "shift out"
+  -- | '\SI' => .some "shift in"
+  -- | '\DLE' => .some "data link escape"
+  -- | '\DC1' => .some "device control one"
+  -- | '\DC2' => .some "device control two"
+  -- | '\DC3' => .some "device control three"
+  -- | '\DC4' => .some "device control four"
+  -- | '\NAK' => .some "negative acknowledge"
+  -- | '\SYN' => .some "synchronous idle"
+  -- | '\ETB' => .some "end of transmission block"
+  -- | '\CAN' => .some "cancel"
+  -- | '\EM' => .some "end of medium"
+  -- | '\SUB' => .some "substitute"
+  -- | '\ESC' => .some "escape"
+  -- | '\FS' => .some "file separator"
+  -- | '\GS' => .some "group separator"
+  -- | '\RS' => .some "record separator"
+  -- | '\US' => .some "unit separator"
+  -- | '\DEL' => .some "delete"
+  -- | '\160' => .some "non-breaking space"
+  | _ => .none
+
+private def charPretty : Char → String
+  | ' ' => "space"
+  | ch  => (charPretty' ch).getD $ s!"'{ch}'"
+
+private def stringPretty : NEList Char → String
+  | ⟦x⟧ => charPretty x
+  | ⟦'\r','\n'⟧ => "crlf newline"
+  | xs =>
+    let f c := match charPretty' c with
+      | .none => s!"{c}"
+      | .some pretty => s!"<{pretty}>"
+    s!"\"{String.join $ f <$> xs.toList}\""
+
+--===========================================================--
+--=================== PRINTABLE INSTANCES ===================--
+--===========================================================--
+
+instance : Printable Char where
```

Comment:
This is a bit awkward because now we can't default to `ToString` or `Repr` representations of bytes.
We probably need to at least cover pairs of things from Zepto for this to be an overall improvement.